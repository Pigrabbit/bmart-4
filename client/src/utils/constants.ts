import { SortType } from '../types/sort'

export const CS_PHONE: string = '1600-0025'
export const CS_EMAIL: string = 'bmart_together@woowahan.com'
export const SLIDER_INTERVAL_TIME = 2000
export const GRAPHQL_URI = process.env.REACT_APP_GRAPHQL_URI
export const SEARCH_URI = process.env.REACT_APP_SEARCH_URI
export const AUTO_SUGGEST_URI = process.env.REACT_APP_AUTO_SUGGEST_URI
export const OAUTH_URI = process.env.REACT_APP_GOOGLE_OAUTH_URI
export const BAEDAL_TIP = 1000
export const MIN_ORDER_PRICE = 5000
export const SKELETON_CARD_COUNT = 10
export const CATEGORIES = [
  {
    id: 0,
    name: '채소',
    displayName: '채소',
    thumbnail:
      'https://thumbnail9.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2020/05/07/17/1/16384cc3-f4d3-4652-9e5f-20064e610673.jpg',
  },
  {
    id: 1,
    name: '과일',
    displayName: '과일',
    thumbnail:
      'https://thumbnail7.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2020/07/09/17/9/a18e621e-ac47-433c-975e-223b2dea2727.jpg',
  },
  {
    id: 2,
    name: '생수/음료',
    displayName: '음료',
    thumbnail:
      'https://thumbnail7.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2019/02/19/16/8/ccd31354-425f-4f1f-b2cd-1a96f88e63d0.jpg',
  },
  {
    id: 3,
    name: '가루/조미료/오일',
    displayName: '조미료',
    thumbnail:
      'https://thumbnail8.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2018/12/20/10/5/843881a6-11b8-4eaa-a147-16130f80110e.jpg',
  },
  {
    id: 4,
    name: '유제품/아이스크림',
    displayName: '유제품',
    thumbnail:
      'https://thumbnail9.coupangcdn.com/thumbnails/remote/230x230ex/image/product/image/vendoritem/2018/10/26/3822437999/837dc477-f18b-4485-98a5-b69caaf21938.jpg',
  },
  {
    id: 5,
    name: '축산/계란',
    displayName: '축산/계란',
    thumbnail:
      'https://thumbnail10.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/86412774225188-82d1d44a-b4e5-4daa-bcf3-ff264898f476.jpg',
  },
  {
    id: 6,
    name: '수산물/건어물',
    displayName: '수산물',
    thumbnail:
      'https://thumbnail7.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2019/02/22/15/5/d364c3a0-0e9a-46dd-b173-5b826ab5db5a.jpg',
  },
  {
    id: 7,
    name: '냉장/냉동/간편요리',
    displayName: '냉장/냉동',
    thumbnail:
      'https://thumbnail6.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2019/09/02/11/4/8e985aeb-5a4c-4159-b749-04b085dd3376.jpg',
  },
  {
    id: 8,
    name: '반찬/간편식/대용식',
    displayName: '반찬/간편식',
    thumbnail:
      'https://thumbnail10.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/12019442178144-5859193f-c48c-43e8-9b2f-033ebac1e25d.jpg',
  },
  {
    id: 9,
    name: '장/소스/드레싱/식초',
    displayName: '장/소스',
    thumbnail: 'https://img-cf.kurly.com/shop/data/goodsview/20181212/gv20000038467_1.jpg',
  },
  {
    id: 10,
    name: '과자/초콜릿/시리얼',
    displayName: '과자/초콜릿',
    thumbnail:
      'https://thumbnail7.coupangcdn.com/thumbnails/remote/230x230ex/image/retail/images/2019/12/17/16/9/68c48bf2-23dc-4a2a-941d-47550237da60.jpg',
  },
]

export const NAVIGATIONS = [
  { name: 'home', displayName: '홈', path: '/' },
  { name: 'search', displayName: '검색', path: '/search' },
  { name: 'favorite', displayName: '찜', path: '/favorite' },
  { name: 'history', displayName: '주문내역', path: '/history' },
  { name: 'cart', displayName: '장바구니', path: '/cart' },
]

export const PRODUCT_SORT_TYPE: { id: SortType; name: string }[] = [
  { id: '', name: '기본 정렬순' },
  { id: 'priceAsc', name: '가격 낮은순' },
  { id: 'priceDesc', name: '가격 높은순' },
]

export const KR_WEEKDAY = ['일', '월', '화', '수', '목', '금', '토']
// relate to CategoryListSection component's lazyloading implements
// 현재 포커싱된 카테고리 이후 몇 개의 카테고리의 loading을 진행할 것인가에 대한 상수
export const CATEGORY_SECTION_LAZYLOAD_ADDER = 2
export const DISCOUNT_PERCENTAGE_CARD_LIMIT = 30

export const CAROUSEL_BASIC_INTERSECTION_RATIO_THRESHOLD = 0.9
export const CAROUSEL_BASIC_INTERSECTION_THRESHOLD = 0.9
export const MAX_PRODUCT_PURCHASE_LIMIT = 10
export const MIN_PRODUCT_PURCHASE_LIMIT = 1
export const ONE_PAGE_LENGTH = 10
export const SPECIAL_CHAR_REGEX = /[~!@#$%^&*()_+|<>?:{}]/

export const MAX_SEARCH_QUERY_LENGTH = 30
export const TUNG_MESSAGE = {
  EMPTY_CART: '장바구니가 텅 비어있어요',
  EMPTY_HISTORY: '주문 내역이 없어요ㅠ',
}
