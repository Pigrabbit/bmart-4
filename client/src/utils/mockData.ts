import { BannerType } from '../types/banner'

export const bigBannerList: BannerType[] = [
  { src: `${process.env.PUBLIC_URL}/images/big/1.jpg`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/big/2.jpg`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/big/3.gif`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/big/4.jpg`, href: 'https://www.google.com' },
]

export const smallBannerList: BannerType[] = [
  { src: `${process.env.PUBLIC_URL}/images/small/1.jpg`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/small/2.jpg`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/small/3.jpg`, href: 'https://www.google.com' },
  { src: `${process.env.PUBLIC_URL}/images/small/4.jpg`, href: 'https://www.google.com' },
]

export const productBannerList: BannerType[] = [
  {
    src:
      'http://thumbnail7.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/502068498864282-291147e0-a12f-4f0e-9099-6171de534165.jpg',
    href: 'https://www.google.com',
  },
  {
    src:
      'http://thumbnail9.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/503436597885335-49a9a2f4-db95-4c12-8571-2a34ad9888f4.jpg',
    href: 'https://www.google.com',
  },
  {
    src:
      'http://thumbnail7.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/503436408371913-a10b8d55-23aa-4e1d-ba7d-ab44edf19547.jpg',
    href: 'https://www.google.com',
  },
  {
    src:
      'http://thumbnail9.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/503436384624178-c61fd94b-3b1f-4cb1-b34a-77ede5b342e0.jpg',
    href: 'https://www.google.com',
  },
  {
    src:
      'http://thumbnail9.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/502068638132244-7e73d69f-b1fa-429e-9879-4fa62d2ac786.jpg',
    href: 'https://www.google.com',
  },
  {
    src:
      'http://thumbnail8.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/502219012342103-de9a5162-5efc-456d-850e-d4e16ee3e7f6.png',
    href: 'https://www.google.com',
  },
]
