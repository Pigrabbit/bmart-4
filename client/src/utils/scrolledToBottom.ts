export const scrolledToBottom = () => {
  let scrollHeight = Math.max(document.documentElement.scrollHeight, document.body.scrollHeight)
  let scrollTop = Math.max(document.documentElement.scrollTop, document.body.scrollTop)
  let clientHeight = document.documentElement.clientHeight
  if (scrollTop + clientHeight === scrollHeight) {
    return true
  } else return false
}
