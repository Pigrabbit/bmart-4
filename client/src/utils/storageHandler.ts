const jwt = require('jsonwebtoken')

export const parseCookieToJson = (cookie: string) => {
  return Object.fromEntries(cookie.split('; ').map((x) => x.split('=')))
}

export const clearCookie = () => {
  document.cookie = `token=;expires=Thu, 01 Jan 1970 00:00:00 GMT;path=/`
}

export const decodeToken = (token: string) => {
  return jwt.decode(token)
}

export const saveJsonToLocalStorage = (jsonObj: any) => {
  Object.keys(jsonObj).map((key) => {
    localStorage.setItem(key, jsonObj[key])
  })
}
