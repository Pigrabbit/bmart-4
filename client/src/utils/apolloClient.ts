import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import { GRAPHQL_URI } from './constants'

const httplink = createHttpLink({
  uri: GRAPHQL_URI,
})
const authLink = setContext((_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = localStorage.getItem('token')
  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  }
})
const cache = new InMemoryCache()
export const client = new ApolloClient({
  link: authLink.concat(httplink),
  cache,
})
