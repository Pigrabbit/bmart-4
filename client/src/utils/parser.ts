import { KR_WEEKDAY } from "./constants"

export const parseToLocalMoneyString = (money: number): string => {
  return String(money).replace(/(\d)(?=(?:\d{3})+(?!\d))/g, '$1,')
}

export const replaceSlashesWithHyphens = (str: String) => {
  return str.replace(/\//g, '-')
}

export const replaceHyphensWithSlashes = (str: String) => {
  return str.replace(/-/g, '/')
}

export const replaceHyphensWithCommas = (str: String) => {
  return str.replace(/-/g, ', ')
}

export const replaceSlashesWithCommas = (str: String) => {
  return str.replace(/\//g, ', ')
}

export const toLocalDateString = (isoString: string) => {
  const dateObj = new Date(isoString)
  const month = dateObj.getMonth() + 1
  const weekday = KR_WEEKDAY[dateObj.getDay()]
  const date = dateObj.getDate()
  return `${month}/${date} (${weekday})`
}