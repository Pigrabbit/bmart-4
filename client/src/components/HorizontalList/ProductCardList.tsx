import React from 'react'
import styled from 'styled-components'
import { ProductCard } from '../ProductCard'
import { ProductCardType } from '../../types/productCard'
import { SkeletonCard } from '../SkeletonCard'
import { SKELETON_CARD_COUNT } from '../../utils/constants'

type Props = {
  loading: boolean
  productList: ProductCardType[]
}

const StyledProductList = styled.div`
  display: flex;
  margin-top: 10px;
`

const StyledSpacer = styled.div`
  width: 3px;
  flex: 0 0 auto;
`

export const ProductCardList = (props: Props) => {
  const { productList, loading } = props

  return (
    <StyledProductList className="product-list">
      {loading
        ? Array(SKELETON_CARD_COUNT)
            .fill('')
            .map((_, idx) => <SkeletonCard key={idx} width="36%" style={{ marginRight: '10px' }} />)
        : productList.map((product, idx) => (
            <ProductCard key={idx} product={product} width="36%" style={{ marginRight: '10px' }} />
          ))}
      <StyledSpacer></StyledSpacer>
    </StyledProductList>
  )
}
