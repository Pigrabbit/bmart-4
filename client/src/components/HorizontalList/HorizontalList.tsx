import React from 'react'
import styled from 'styled-components'
import { ProductCardList } from './ProductCardList'
import { STYLES } from '../../utils/styleConstants'
import { StyledWrapper } from '../../styles/StyledWrapper'
import { ProductCardType } from '../../types/productCard'

export type Props = {
  title: string
  productList: ProductCardType[]
  double?: boolean
  loading?: boolean
}

const StyledContainer = styled.div`
  padding: 12px 0;
  width: 100%;
  overflow-x: hidden;
`
const StyledHeader = styled.div`
  padding-left: ${STYLES.padding};
  h2 {
    margin: 0;
    line-height: 30px;
  }
`
const StyledProductListWrap = styled.div`
  overflow-x: auto;
  padding-left: ${STYLES.padding};
  scroll-behavior: smooth;
  -webkit-overflow-scrolling: touch;
`

export const HorizontalList = (props: Props) => {
  const { productList, double = false, loading = false } = props

  let { title } = props

  const renderProductCardList = () => {
    const length = productList.length
    const half = Math.ceil(length / 2)

    return double ? (
      <>
        <ProductCardList loading={loading} productList={[...productList.slice(0, half)]} />
        <ProductCardList loading={loading} productList={[...productList.slice(half, length)]} />
      </>
    ) : (
      <ProductCardList loading={loading} productList={productList} />
    )
  }

  const firstname: string | null = localStorage.getItem('firstname')
  if (firstname) {
    title = title.replace('회원', firstname)
  }

  return (
    <StyledWrapper className="horizontal-product-list">
      <StyledContainer>
        <StyledHeader>
          <h2 className="title">{title}</h2>
        </StyledHeader>
        <StyledProductListWrap>{renderProductCardList()}</StyledProductListWrap>
      </StyledContainer>
    </StyledWrapper>
  )
}
