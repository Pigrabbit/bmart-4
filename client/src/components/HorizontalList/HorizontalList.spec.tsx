import React from 'react'
import { HorizontalList } from './HorizontalList'
import { mount } from 'enzyme'
import { productList } from '../../utils/mockData'
import { MemoryRouter } from 'react-router-dom'

import { Props } from './HorizontalList'

const testProps: Props = {
  title: '테스트',
  double: false,
  productList,
}

describe('<HorizontalList />', () => {
  const wrapper = mount(
    <MemoryRouter>
      <HorizontalList {...testProps} />
    </MemoryRouter>
  )

  it('match: snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('render: double prop에 따라 올바른 상품 리스트를 노출하는가', () => {
    // double: false 일 때 리스트 1개, 상품갯수는 props로 전달한 갯수만큼
    expect(wrapper.find('.product-list').hostNodes().length).toStrictEqual(1)
    expect(wrapper.find('.product-card').hostNodes().length).toStrictEqual(
      testProps.productList.length
    )

    wrapper.setProps(
      {
        children: <HorizontalList {...{ ...testProps, double: true }} />,
      },
      () => {
        // double: true 일 때 리스트 2개, 상품갯수는 props로 전달한 갯수만큼 (위와 동일)
        expect(wrapper.find('.product-list').hostNodes().length).toStrictEqual(2)
        expect(wrapper.find('.product-card').hostNodes().length).toStrictEqual(
          testProps.productList.length
        )
      }
    )
  })
})
