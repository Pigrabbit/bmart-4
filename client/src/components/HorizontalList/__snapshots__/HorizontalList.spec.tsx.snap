// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HorizontalList /> match: snapshot 1`] = `ReactWrapper {}`;
